From a8ef9249c37c63452d57662b2f69e33a35175073 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:10:09 -0400
Subject: [powerpc] book3s: Fix CFAR clobbering issue in machine check handler

Message-id: <1410545655-205645-381-git-send-email-dzickus@redhat.com>
Patchwork-id: 94358
O-Subject: [RHEL7 PATCH 380/626] powerpc/book3s: Fix CFAR clobbering issue in machine check handler.
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit d410ae2126481a74bc5be4a6242c4a232c19a984
Author: Mahesh Salgaonkar <mahesh@linux.vnet.ibm.com>
Date:   Tue Mar 11 10:56:18 2014 +0530

    powerpc/book3s: Fix CFAR clobbering issue in machine check handler.

    While checking powersaving mode in machine check handler at 0x200, we
    clobber CFAR register. Fix it by saving and restoring it during beq/bgt.

    Signed-off-by: Mahesh Salgaonkar <mahesh@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/exception-64s.h b/arch/powerpc/include/asm/exception-64s.h
index 6683061..aeaa56c 100644
--- a/arch/powerpc/include/asm/exception-64s.h
+++ b/arch/powerpc/include/asm/exception-64s.h
@@ -147,6 +147,14 @@ BEGIN_FTR_SECTION_NESTED(943)      \
 END_FTR_SECTION_NESTED(ftr,ftr,943)
 
 /*
+ * Set an SPR from a register if the CPU has the given feature
+ */
+#define OPT_SET_SPR(ra, spr, ftr)     \
+BEGIN_FTR_SECTION_NESTED(943)      \
+ mtspr spr,ra;       \
+END_FTR_SECTION_NESTED(ftr,ftr,943)
+
+/*
  * Save a register to the PACA if the CPU has the given feature
  */
 #define OPT_SAVE_REG_TO_PACA(offset, ra, ftr)    \
diff --git a/arch/powerpc/kernel/exceptions-64s.S b/arch/powerpc/kernel/exceptions-64s.S
index 9533d7a..d9c650e 100644
--- a/arch/powerpc/kernel/exceptions-64s.S
+++ b/arch/powerpc/kernel/exceptions-64s.S
@@ -170,13 +170,18 @@ BEGIN_FTR_SECTION
   */
  mfspr r13,SPRN_SRR1
  rlwinm. r13,r13,47-31,30,31
+ OPT_GET_SPR(r13, SPRN_CFAR, CPU_FTR_CFAR)
  beq 9f
 
+ mfspr r13,SPRN_SRR1
+ rlwinm. r13,r13,47-31,30,31
  /* waking up from powersave (nap) state */
  cmpwi cr1,r13,2
  /* Total loss of HV state is fatal. let's just stay stuck here */
+ OPT_GET_SPR(r13, SPRN_CFAR, CPU_FTR_CFAR)
  bgt cr1,.
 9:
+ OPT_SET_SPR(r13, SPRN_CFAR, CPU_FTR_CFAR)
 END_FTR_SECTION_IFSET(CPU_FTR_HVMODE | CPU_FTR_ARCH_206)
 #endif /* CONFIG_PPC_P7_NAP */
  EXCEPTION_PROLOG_0(PACA_EXMC)
-- 
1.7.1